packages/gatsby/src/commands/build.ts

```diff
@@ -201,6 +202,20 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
     buildSSRBundleActivityProgress.end()
   }
 
+  if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+    const validateEnginesActivity = report.activityTimer(`Validating engines`, {
+      parentSpan: buildSpan,
+    })
+    validateEnginesActivity.start()
+    try {
+      await validateEngines(store.getState().program.directory)
+    } catch (err) {
+      validateEnginesActivity.panic(err)
+    } finally {
+      validateEnginesActivity.end()
+    }
+  }
+
```

Comment:
Note that this happens after engines are generated and after build-html is generated. Ideally validation happen as part of building engines activity, but as it is right now it would fail validation just because `routes/render-page` (etc) are not generated yet, so imports fail.

As alternative I can move it up and instead just "mock" `render-page` in our validator to not fail on those (but I don't think this is good idea because this just disables part of code (render-page) from same validation as rest of engines have